sumpy/expansion/multipole.py

```diff
@@ -166,67 +166,129 @@ def translate_from(self, src_expansion, src_coeff_exprs, src_rscale,
         tgt_mi_to_index = {mi: i for i, mi in enumerate(
             self.get_full_coefficient_identifiers())}
 
-        src_coeff_exprs = list(src_coeff_exprs)
-        for i, mi in enumerate(src_expansion.get_coefficient_identifiers()):
-            src_coeff_exprs[i] *= sym.UnevaluatedExpr(src_rscale/tgt_rscale)**sum(mi)
-
-        result = [0] * len(self.get_full_coefficient_identifiers())
-
         # This algorithm uses the observation that M2M coefficients
         # have the following form in 2D
         #
-        # $B_{m, n} = \sum_{i\le m, j\le n} A_{i, j}
+        # $T_{m, n} = \sum_{i\le m, j\le n} C_{i, j}
         #             d_x^i d_y^j \binom{m}{i} \binom{n}{j}$
         # and can be rewritten as follows.
         #
-        # Let $T_{m, n} = \sum_{i\le m} A_{i, n} d_x^i \binom{m}{i}$.
+        # Let $Y_{m, n} = \sum_{i\le m} C_{i, n} d_x^i \binom{m}{i}$.
         #
-        # Then, $B_{m, n} = \sum_{j\le n} T_{m, j} d_y^j \binom{n}{j}$.
+        # Then, $T_{m, n} = \sum_{j\le n} Y_{m, j} d_y^j \binom{n}{j}$.
         #
-        # $T_{m, n}$ are $p^2$ number of temporary variables that are
+        # $Y_{m, n}$ are $p^2$ number of temporary variables that are
         # reused for different M2M coefficients and costs $p$ per variable.
-        # Total cost for calculating $T_{m, n}$ is $p^3$ and similar
-        # for $B_{m, n}$
+        # Total cost for calculating $Y_{m, n}$ is $p^3$ and similar
+        # for $T_{m, n}$.
 
         # In other words, we're better off computing the translation
-        # one dimension at a time. This is realized here by using
-        # the output from one dimension as the input to the next.
-        # per_dim_coeffs_to_translate serves as the array of input
-        # coefficients for each dimension's translation.
-
-        dim_coeffs_to_translate = src_coeff_exprs
+        # one dimension at a time. If the coefficients in the source
+        # expansion have the form (0, m) and (n, 0), where m>=0, n>=1,
+        # then we calculate the output from (0, m) with the second
+        # dimension as the fastest varying dimension and then calculate
+        # the output from (n, 0) with the first dimension as the fastest
+        # varying dimension.
+
+        # Note that for calculating Y values, C is the input
+        # and for calculating the final result Y is the input.
+
+        tgt_split = \
+            self.expansion_terms_wrangler._get_coeff_identifier_split()
+        result = [0] * len(self.get_full_coefficient_identifiers())
 
-        mi_to_index = src_mi_to_index
-        for d in range(self.dim):
+        # For this algorithm, we need the hyperplanes that are parallel
+        # to each other grouped.
+        non_zero_coeffs_grouped_by_normal_dir = [[] for d in range(self.dim)]
+        for d, tgt_mi in tgt_split:
+            non_zero_coeffs_grouped_by_normal_dir[d] += tgt_mi
+
+        for normal_dir in set(d for d, _ in tgt_split):
```

Comment:
These two loops could probably be fused, avoiding the need for the `non_zero_coeffs_grouped_by_normal_dir` temporary.